src/logger.cc

```diff
@@ -9,17 +9,37 @@ const char kLogInfo[] = "ninja: ";
 const char kLogWarning[] = "ninja: warning: ";
```

Comment:
I recommend taking a look at the Boost logging framework, which handles this very differently by allowing you to have multiple different logger destinations addressed by parameters to the logger being written to

Reply:
I'm fine looking in to that, I was under the impression that the maintainers of ninja would strongly prefer not to add any large new dependencies to the project. I was shooting for something relatively simple to implement that makes the default output behavior for ninja the same.

Reply:
I wasn't implying that you should directly use the Boost logging framework.
Just to look at it as a source of inspiration.

Reply:
I've never used the boost logging framework before, but I've now read some of the documentation after your suggestion. The logging framework is quite large, complex, and configurable. That's neat, but I don't think we want to build something like that entirely for ninja. Could you discuss one or two high points from boost logging you think I should incoporate?

Reply:
The most important aspects of it for the code that I'm writing is the following:

1. Multiple backends, so I can dynamically target log files, console output, or a logging server on the network, as I feel like at runtime.
2. Dynamic variables captured with each log statement, which can include the current functions name, or the current timestamp, or what have you. 
3. Easy to replace the functionality using macros. E.g. I can provide my own logging statements that customize what gets logged, by defining a macro that i call.

Reply:
Thanks, that's helpful. That sounds like an enormous amount of work considering that all I originally wanted from the logging system was a way for library users to capture messages themselves and everyone else to continue to get information on stdout. If you were writing this how long do you think it would take you?

Reply:
Honestly, if I were doing it for the Ninja project, I would just define some macro, e.g. NINJA_LOG_STREAM, or what have you, and only define the macro to the ninja project's preferred default if it doesn't already have a definition.

That allows consumers of the Ninja project to inject their own log implementation if they want to.

Reply:
You could get a little bit more robust with

#define NINJA_LOG_STREAM(Severity) /**/

to allow for custom log implementations to capture the severity identifier as well, but it's not strictly necessary.

Reply:
Two completely different way that this could be handled, if using macros isn't the right approach.

Dynamic polymorphism: Have Ninja provide an abstract base class type (e.g. an Interface) that can be inherited from, and then have the Ninja library accept an instance of this type when the library is instantiated. Any logging calls happen through this interface, using virtual functions.

Static polymorphism: Have the ninja codebase accept the logging object to use as a template parameter. This requires significantly more changes to the ninja codebase than the Dynamic polymorphism version though.

Implementing Dynamic polymorphism is a step on the road to static polymorphism regardless, so I'd start with that and then if the performance impact of the virtual function calls is too much,  investigate the static approach.

Reply:
If the API for either version of the logging concepts above accept a `char (&)[]` (reference to character array) and length, we can use the __PRETTY_FUNCTION__ identifier (or similar) to still log the function name that's calling the logging code.